frontend/src/metabase-lib/moving-average.unit.spec.ts

```diff
@@ -0,0 +1,1132 @@
+import { ORDERS_ID } from "metabase-types/api/mocks/presets";
+
+import { aggregate, aggregations } from "./aggregation";
+import { displayInfo } from "./metadata";
+import {
+  diffMovingAverageClause,
+  movingAverageClause,
+  percentDiffMovingAverageClause,
+} from "./moving-average";
+import { toLegacyQuery } from "./query";
+import { SAMPLE_DATABASE, createQueryWithClauses } from "./test-helpers";
+import type { Query } from "./types";
+
+const stageIndex = -1;
+
+const queryNoBreakout = createQueryWithClauses({
+  aggregations: [{ operatorName: "count" }],
+});
+
+const queryDateBreakoutNoBinning = createQueryWithClauses({
+  query: queryNoBreakout,
+  breakouts: [
+    {
+      columnName: "CREATED_AT",
+      tableName: "ORDERS",
+    },
+  ],
+});
+
+const queryDateBreakoutBinning = createQueryWithClauses({
+  query: queryNoBreakout,
+  breakouts: [
+    {
+      columnName: "CREATED_AT",
+      tableName: "ORDERS",
+      temporalBucketName: "Month",
+    },
+  ],
+});
+
+const queryCategoryBreakout = createQueryWithClauses({
+  query: queryNoBreakout,
+  breakouts: [
+    {
+      columnName: "CATEGORY",
+      tableName: "PRODUCTS",
+    },
+  ],
+});
+
+describe("movingAverageClause", () => {
+  const setup = (
+    query: Query,
+    offset: number,
+    includeCurrentPeriod: boolean = false,
+  ) => {
+    const [aggregation] = aggregations(query, stageIndex);
+    const clause = movingAverageClause(
+      query,
+      stageIndex,
+      aggregation,
+      offset,
+      includeCurrentPeriod,
+    );
+    const finalQuery = aggregate(query, stageIndex, clause);
+
+    return {
+      clause,
+      query: finalQuery,
+    };
+  };
+
+  describe("includeCurrentPeriod = false", () => {
+    const includeCurrentPeriod = false;
+
+    describe("offset = -2", () => {
+      const offset = -2;
+
+      describe("no breakout", () => {
+        const { query, clause } = setup(
+          queryNoBreakout,
+          offset,
+          includeCurrentPeriod,
+        );
+        it("produces correct aggregation name", () => {
+          const info = displayInfo(query, stageIndex, clause);
+          expect(info.displayName).toBe("Count (2-period moving average)");
+        });
+
+        it("produces correct aggregation clause", () => {
+          expect(toLegacyQuery(query)).toMatchObject({
```

Comment:
Any direct MBQL should be avoided. Please use `expressionParts` with `expect(...).toMatchObject`